fmn/tests/rules/test_greenwave.py

```diff
@@ -0,0 +1,110 @@
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Suite 500, Boston, MA 02110-1335,  USA
+"""Tests for the :mod:`fmn.rules.greenwave` module."""
+from __future__ import unicode_literals
+
+import copy
+
+from fmn.rules import greenwave
+from fmn.tests import Base
+
+
+GREENWAVE_MSG = {
+    "username": "openshift",
+    "i": 59,
+    "timestamp": 1508934768.0,
+    "msg_id": "2017-db45b07c-b39a-4fd3-af49-ae7cb83bb78b",
+    "crypto": "x509",
+    "topic": "org.fedoraproject.prod.greenwave.greenwave.decision.update",
+    "msg": {
+        "policies_satisified": False,
+        "decision_context": "bodhi_update_push_stable",
+        "product_version": "fedora-27",
+        "applicable_policies": [
+            "taskotron_release_critical_tasks_for_stable",
+            "taskotron_release_critical_tasks_for_stable_with_blacklist"
+        ],
+        "unsatisfied_requirements": [
+            {
+                "testcase": "dist.abicheck",
+                "item": {
+                    "item": "lilypond-doc-2.19.80-1.fc25",
+                    "type": "koji_build"
+                },
+                "type": "test-result-missing"
+            }
+        ],
+        "subject": [
+            {
+                "item": "lilypond-doc-2.19.80-1.fc25",
+                "type": "koji_build"
+            }
+        ],
+        "summary": "1 of 3 required tests not found",
+        "previous": {
+            "policies_satisified": False,
+            "summary": "2 of 3 required tests not found",
+            "unsatisfied_requirements": [
+                {
+                    "testcase": "dist.upgradepath",
+                    "item": {
+                        "item": "lilypond-doc-2.19.80-1.fc25",
+                        "type": "koji_build"
+                    },
+                    "type": "test-result-missing"
+                },
+                {
+                    "testcase": "dist.abicheck",
+                    "item": {
+                        "item": "lilypond-doc-2.19.80-1.fc25",
+                        "type": "koji_build"
+                    },
+                    "type": "test-result-missing"
+                }
+            ],
+            "applicable_policies": [
+                "taskotron_release_critical_tasks_for_stable",
+                "taskotron_release_critical_tasks_for_stable_with_blacklist"
+            ]
+        }
+    }
+}
+
+
+class TestGreenwaveDecisionUpdate(Base):
+    '''Test greenwave_decision_update()'''
+
+    def setUp(self):
+        super(TestGreenwaveDecisionUpdate, self).setUp()
+        self.msg = copy.deepcopy(GREENWAVE_MSG)
```

Comment:
I don't think any rule should ever mutate the message so I'd recommend dropping this. This way if a rule _does_ mutate the message tests will fail.

Reply:
> I don't think any rule should ever mutate the message so I'd recommend dropping this. This way if a rule does mutate the message tests will fail.

Ah, but some of the *tests* here do mutate the message.